File: app/src/main/java/com/hfad/mymovies/ui/DetailFragment.java
package com.hfad.mymovies.ui;

import android.os.Bundle;
import android.util.Log;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.fragment.app.Fragment;
import androidx.lifecycle.Observer;
import androidx.lifecycle.ViewModelProvider;
import androidx.navigation.Navigation;

import com.hfad.mymovies.adapters.CastAdapter;
import com.hfad.mymovies.adapters.MovieAdapter;
import com.hfad.mymovies.adapters.ReviewAdapter;
import com.hfad.mymovies.data.models.Actors;
import com.hfad.mymovies.data.models.Movie;
import com.hfad.mymovies.data.models.MovieDiscover;
import com.hfad.mymovies.data.models.Review;
import com.hfad.mymovies.databinding.FragmentDetailBinding;
import com.hfad.mymovies.utils.InjectorUtils;
import com.hfad.mymovies.viewmodels.DetailFragmentViewModel;

import java.util.List;

public class DetailFragment extends Fragment {

    private ReviewAdapter mReviewAdapter;
    private CastAdapter mCastAdapter;
    private FragmentDetailBinding mBinding;
    private DetailFragmentViewModel mViewModel;
    private MovieAdapter mMovieAdapter;

    @Nullable
    @Override
    public View onCreateView(@NonNull LayoutInflater inflater,
                             @Nullable ViewGroup container,
                             @Nullable Bundle savedInstanceState) {

        mBinding = FragmentDetailBinding.inflate(inflater, container, false);
        int movieId = DetailFragmentArgs.fromBundle(getArguments()).getMovieId();

        mViewModel = new ViewModelProvider(this,
                        InjectorUtils.provideDetailFragmentViewModelFactory(getContext(), movieId))
                .get(DetailFragmentViewModel.class);

        mBinding.imageView.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                Navigation.findNavController(view).navigateUp();
            }
        });

        subscribeUi();

        mMovieAdapter = new MovieAdapter(this);
        mReviewAdapter = new ReviewAdapter();
        mCastAdapter = new CastAdapter();
        mBinding.rvRecommendations.setAdapter(mMovieAdapter);
        mBinding.rvReviews.setAdapter(mReviewAdapter);
        mBinding.rvActors.setAdapter(mCastAdapter);

        setCallback();

        return mBinding.getRoot();
    }

    private void setCallback(){
        Callback callback = new Callback(){
            @Override
            public void addOrDelete(Movie movie, Boolean isFavourite) {
                if(isFavourite) {
                    mViewModel.deleteFromFavourite();
                } else {
                    mViewModel.addMovieToFavourite(movie);
                }
            }
        };
        mBinding.setCallback(callback);
    }

    private void loadDetail(boolean isFavourite){
        if(isFavourite){
            mBinding.setFavouriteMovie(mViewModel.getFavouriteMovie());
        } else {
            mViewModel.getMovie().observe(getViewLifecycleOwner(), new Observer<Movie>() {
                @Override
                public void onChanged(Movie movie) {
                    mBinding.setMovie(movie);
                }
            });
        }
    }

    private void subscribeUi(){

        mViewModel.getReviews().observe(getViewLifecycleOwner(), new Observer<List<Review>>() {
            @Override
            public void onChanged(List<Review> reviews) {
                mReviewAdapter.setReviews(reviews);
            }
        });

        mViewModel.getMovies().observe(getViewLifecycleOwner(), new Observer<List<MovieDiscover>>() {
            @Override
            public void onChanged(List<MovieDiscover> movieDiscovers) {
                mMovieAdapter.clear();
                mMovieAdapter.setMovieList(movieDiscovers);
            }
        });

        mViewModel.isFavourite().observe(getViewLifecycleOwner(), new Observer<Boolean>() {
            @Override
            public void onChanged(Boolean aBoolean) {
                loadDetail(aBoolean);
                mBinding.setIsFavourite(aBoolean);
            }
        });

        mViewModel.getActors().observe(getViewLifecycleOwner(), new Observer<List<Actors>>() {
            @Override
            public void onChanged(List<Actors> actors) {
                Log.i("ProblemDF", "" + actors.size());
                mCastAdapter.setActorsList(actors);
            }
        });
    }

    @Override
    public void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);

    }

    public interface Callback {
        void addOrDelete(Movie movie, Boolean isFavourite);
    }
}

File: app/src/main/java/com/hfad/mymovies/ui/DetailFragment.kt
package com.hfad.mymovies.ui

import android.os.Bundle
import android.util.Log
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.fragment.app.Fragment
import androidx.lifecycle.ViewModelProvider
import androidx.navigation.Navigation
import com.hfad.mymovies.Application
import com.hfad.mymovies.MainActivity
import com.hfad.mymovies.adapters.CastAdapter
import com.hfad.mymovies.adapters.MovieAdapter
import com.hfad.mymovies.adapters.ReviewAdapter
import com.hfad.mymovies.data.models.Movie
import com.hfad.mymovies.databinding.FragmentDetailBinding
import com.hfad.mymovies.di.ViewModelFactory
import com.hfad.mymovies.utils.NetworkUtilsConstants
import com.hfad.mymovies.viewmodels.DetailFragmentViewModel
import com.squareup.picasso.Picasso
import javax.inject.Inject

class DetailFragment : Fragment() {

    private var reviewAdapter: ReviewAdapter? = null
    private var castAdapter: CastAdapter? = null
    @Inject
    lateinit var viewModelFactory: ViewModelFactory
    lateinit var viewModel: DetailFragmentViewModel
    private lateinit var binding: FragmentDetailBinding
    private lateinit var movieAdapter: MovieAdapter

    override fun onCreateView(inflater: LayoutInflater,
                              container: ViewGroup?,
                              savedInstanceState: Bundle?): View {
        ((activity as MainActivity).application as Application).appComponentFactory.inject(this)
        binding = FragmentDetailBinding.inflate(inflater, container, false)
        viewModel = ViewModelProvider(this, viewModelFactory)[DetailFragmentViewModel::class.java]
        val movieId = DetailFragmentArgs.fromBundle(requireArguments()).movieId
        viewModel.movieId = movieId
        initAll()
        binding.detailFragmentBackArrowIv.setOnClickListener { view -> Navigation.findNavController(view).navigateUp() }
        return binding.root
    }

    private fun initAll(){
        subscribeUi()
        initAdapters()
        loadDetail(false)
        setCallback()
    }

    private fun initAdapters() {
        movieAdapter = MovieAdapter(this)
        reviewAdapter = ReviewAdapter()
        castAdapter = CastAdapter()
        binding.detailFragmentRecommendationsRv.adapter = movieAdapter
        binding.detailFragmentReviewsRv.adapter = reviewAdapter
        binding.detailFragmentActorsRv.adapter = castAdapter
    }

    private fun setCallback() {
//        val callback: Callback = object : Callback {
//            override fun addOrDelete(movie: Movie?, isFavourite: Boolean) {
//                if (isFavourite) {
//                    viewModel.deleteFromFavourite()
//                } else {
//                    viewModel.addMovieToFavourite(movie)
//                }
//            }
//        }
        //        binding.setCallback(callback);
    }

    private fun loadDetail(isFavourite: Boolean) {
        if (isFavourite) {
//            binding.setFavouriteMovie(mViewModel.getFavouriteMovie());
        } else {
            viewModel.movie.observe(viewLifecycleOwner, { movie ->
                val url = NetworkUtilsConstants.BASE_POSTER_URL + NetworkUtilsConstants.BIG_POSTER_SIZE + movie.posterPath
                Picasso.get().load(url).into(binding.detailFragmentPosterIv)
                binding.detailFragmentTitle2Tv.text = movie.title
                binding.detailFragmentTitleOriginalTv.text = movie.originalTitle
                val stringBuilder = StringBuilder()
                for (i in movie.genres.indices) {
                    stringBuilder.append(movie.genres[i].name)
                    if (i < movie.genres.size - 1) {
                        stringBuilder.append(" - ")
                    }
                }
                binding.detailFragmentGenreTv.text = stringBuilder
                binding.detailFragmentRatingTv.text = movie.voteAverage.toString()
                binding.detailFragmentReleaseDateTv.text = movie.releaseDate
                binding.detailFragmentOverviewTv.text = movie.overview
            })
        }
    }

    private fun subscribeUi() {
        viewModel.reviews.observe(viewLifecycleOwner, { reviews -> reviewAdapter!!.setReviews(reviews) })
        viewModel.movies.observe(viewLifecycleOwner, { movieDiscovers ->
            movieAdapter.clear()
            movieAdapter.setMovieList(movieDiscovers)
        })
//        viewModel.isFavourite.observe(viewLifecycleOwner, { aBoolean ->
//            loadDetail(aBoolean)
//                            binding.setIsFavourite(aBoolean);
//        })
        viewModel.actors.observe(viewLifecycleOwner, { actors ->
            Log.i("ProblemDF", "" + actors.size)
            castAdapter!!.setActorsList(actors)
        })
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
    }

    interface Callback {
        fun addOrDelete(movie: Movie?, isFavourite: Boolean)
    }
}